src/liquidity.rs

```diff
@@ -31,47 +35,126 @@ use std::time::Duration;
+	address: SocketAddress,
+	token: Option<String>,
+	client_config: LSPS1ClientConfig,
+}
+
 struct LSPS2Service {
 	node_id: PublicKey,
 	address: SocketAddress,
 	token: Option<String>,
+	client_config: LSPS2ClientConfig,
 	pending_fee_requests: Mutex<HashMap<RequestId, oneshot::Sender<LSPS2FeeResponse>>>,
 	pending_buy_requests: Mutex<HashMap<RequestId, oneshot::Sender<LSPS2BuyResponse>>>,
 }
 
-pub(crate) struct LiquiditySource<L: Deref>
+pub(crate) struct LiquiditySourceBuilder<L: Deref>
 where
 	L::Target: LdkLogger,
 {
+	lsps1_service: Option<LSPS1Service>,
 	lsps2_service: Option<LSPS2Service>,
 	channel_manager: Arc<ChannelManager>,
 	keys_manager: Arc<KeysManager>,
-	liquidity_manager: Arc<LiquidityManager>,
+	chain_source: Arc<ChainSource>,
 	config: Arc<Config>,
 	logger: L,
 }
 
-impl<L: Deref> LiquiditySource<L>
+impl<L: Deref> LiquiditySourceBuilder<L>
 where
 	L::Target: LdkLogger,
 {
-	pub(crate) fn new_lsps2(
-		node_id: PublicKey, address: SocketAddress, token: Option<String>,
+	pub(crate) fn new(
 		channel_manager: Arc<ChannelManager>, keys_manager: Arc<KeysManager>,
-		liquidity_manager: Arc<LiquidityManager>, config: Arc<Config>, logger: L,
+		chain_source: Arc<ChainSource>, config: Arc<Config>, logger: L,
 	) -> Self {
+		let lsps1_service = None;
+		let lsps2_service = None;
+		Self {
+			lsps1_service,
+			lsps2_service,
+			channel_manager,
+			keys_manager,
+			chain_source,
+			config,
+			logger,
+		}
+	}
+
+	pub(crate) fn lsps1_service(
+		&mut self, node_id: PublicKey, address: SocketAddress, token: Option<String>,
+	) -> &mut Self {
+		// TODO: allow to set max_channel_fees_msat
+		let client_config = LSPS1ClientConfig { max_channel_fees_msat: None };
+		self.lsps1_service = Some(LSPS1Service { node_id, address, token, client_config });
+		self
+	}
+
+	pub(crate) fn lsps2_service(
+		&mut self, node_id: PublicKey, address: SocketAddress, token: Option<String>,
+	) -> &mut Self {
+		let client_config = LSPS2ClientConfig {};
 		let pending_fee_requests = Mutex::new(HashMap::new());
 		let pending_buy_requests = Mutex::new(HashMap::new());
-		let lsps2_service = Some(LSPS2Service {
+		self.lsps2_service = Some(LSPS2Service {
 			node_id,
 			address,
 			token,
+			client_config,
 			pending_fee_requests,
 			pending_buy_requests,
 		});
-		Self { lsps2_service, channel_manager, keys_manager, liquidity_manager, config, logger }
+		self
 	}
 
+	pub(crate) fn build(self) -> LiquiditySource<L> {
+		let lsps1_client_config = self.lsps1_service.as_ref().map(|s| s.client_config.clone());
+		let lsps2_client_config = self.lsps2_service.as_ref().map(|s| s.client_config.clone());
+		let liquidity_client_config =
+			Some(LiquidityClientConfig { lsps1_client_config, lsps2_client_config });
+
+		let liquidity_manager = Arc::new(LiquidityManager::new(
+			Arc::clone(&self.keys_manager),
+			Arc::clone(&self.channel_manager),
+			Some(Arc::clone(&self.chain_source)),
+			None,
+			None,
+			liquidity_client_config,
```

Comment:
Why do both the `LiquidityManager` manager and each LSPS service need a copy of the config?

Reply:
Oh, I meant `LSPS1ClientConfig` and `LSPS2ClientConfig2`. Each config is cloned from the respective LSPS service, but the originals are never used anywhere else, AFAICT. Just wondering why we hang on to those inside each service. Maybe instead the config should be pulled out of the service and have the builder hold `Option<(LSPS1Service, LSPS1ClientConfig)>` if the config is not needed later. Then you won't have two copies of each config.

Reply:
SGTM. Ideally, we don't have multiple copies so they don't go out-of-sync if they somehow can change in the future.